bluesky/callbacks/zmq.py

```diff
@@ -0,0 +1,257 @@
+    """
+    Dispatch documents received over the network from a 0MQ proxy.
+
+    Parameters
+    ----------
+    address : tuple
+        Address of a running 0MQ proxy, given either as a string like
+        ``'127.0.0.1:5567'`` or as a tuple like ``('127.0.0.1', 5567)``
+    hostname : string, optional
+        A filter: only process documents from a host with this name.
+    pid : int, optional
+        A filter: only process documents from a process with this pid.
+    run_engine_id : int, optional
+        A filter: only process documents from a RunEngine with this Python id
+        (memory address).
+    loop : zmq.asyncio.ZMQEventLoop, optional
+    zmq : object, optional
+        By default, the 'zmq' module is imported and used. Anything else
+        mocking its interface is accepted.
+    zmq_asyncio : object, optional
+        By default, the 'zmq.asyncio' module is imported and used. Anything
+        else mocking its interface is accepted.
+
+    Example
+    -------
+
+    Print all documents generated by remote RunEngines.
+
+    >>> d = RemoteDispatcher(('localhost', 5568))
+    >>> d.subscribe('all', print)
+    >>> d.start()  # runs until interrupted
+    """
+    def __init__(self, address, *, hostname=None, pid=None, run_engine_id=None,
+                 loop=None, zmq=None, zmq_asyncio=None):
+        if zmq is None:
+            import zmq
+        if zmq_asyncio is None:
+            import zmq.asyncio as zmq_asyncio
+        if isinstance(address, str):
+            address = address.split(':')
+        self.address = (address[0], int(address[1]))
+        self.hostname = hostname
+        self.pid = pid
+        self.run_engine_id = run_engine_id
+
+        if loop is None:
+            loop = zmq_asyncio.ZMQEventLoop()
+        self._loop = loop
+        asyncio.set_event_loop(self._loop)
+        self._context = zmq_asyncio.Context()
+        self._socket = self._context.socket(zmq.SUB)
+        url = "tcp://%s:%d" % self.address
+        self._socket.connect(url)
+        self._socket.setsockopt_string(zmq.SUBSCRIBE, "")
+        self._task = None
+
+        def is_our_message(_hostname, _pid, _RE_id):
+            # Close over filters and decide if this message applies to this
+            # RemoteDispatcher.
+            return ((hostname is None or hostname == _hostname)
+                    and (pid is None or pid == _pid)
+                    and (run_engine_id is None or run_engine_id ==
+                         run_engine_id))
+        self._is_our_message = is_our_message
+
+        super().__init__()
+
+    @asyncio.coroutine
+    def _poll(self):
+        while True:
+            message = yield from self._socket.recv()
+            hostname, pid, RE_id, name, doc = message.decode().split(' ', 4)
+            if self._is_our_message(hostname, pid, RE_id):
+                doc = ast.literal_eval(doc)
+                self._loop.call_soon(self.process, DocumentNames[name], doc)
+
+    def start(self):
```

Comment:
Slight preference for the pattern we are using in the caproto servers of these object exposing a `async def run` and have `stop` poison that loop.